File: python/src/sort/strand_sort.py


def strand_sort( in_array ):
    result=[]
    while len(in_array):
        sublist = loop([in_array.pop(0)], in_array)
        result  = merge(result,sublist)
    
    return result

def loop(sublist, in_array):
    i=0
    while i < len(in_array):
        if in_array[i] > sublist[-1]:
            sublist += [ in_array.pop(i) ]
        else:
            i += 1
    return sublist
        
    
def merge(left, right):
    result = []
    while len(left) and len(right):
        if left[0] < right[0]:
            result += [left.pop(0)]
        else:
            result += [right.pop(0)]
    return result + left + right

File: c#/Sort/StrandSort.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Sort
{
    public class StrandSort : ISort
    {

        public ArrayList Sort(ArrayList inArray)
        {
            var result = new ArrayList();
            while(inArray.Count > 0)
            {
                var subList = Loop(ref inArray);
                result = Merge(result, subList);
            }
            
            return result;
        }

        private ArrayList Merge(ArrayList left, ArrayList right)
        {
            return MergeSort.Merge(left, right);
        }


        private ArrayList Loop(ref ArrayList inArray)
        {
            var subList = new ArrayList { inArray[0] };
            inArray.RemoveAt(0);

            int i = 0;
            while( i < inArray.Count)
            {
                if ((int)inArray[i] > (int)subList[subList.Count-1])
                {
                    subList.Add(inArray[i]);
                    inArray.RemoveAt(i);
                }
                else
                    i++;
                     
            }

            return subList;
        }
    }
}
